src/csharp/Grpc.Core/Marshaller.cs

```diff
@@ -29,36 +29,122 @@ public class Marshaller<T>
+        public Func<byte[], T> Deserializer => this.deserializer;
+
+        /// <summary>
+        /// Gets the serializer function.
+        /// Note: experimental API that can change or be removed without any prior notice.
+        /// </summary>
+        public Action<T, SerializationContext> ContextualSerializer => this.contextualSerializer;
+
+        /// <summary>
+        /// Gets the serializer function.
+        /// Note: experimental API that can change or be removed without any prior notice.
+        /// </summary>
+        public Func<DeserializationContext, T> ContextualDeserializer => this.contextualDeserializer;
+
+        // for backward compatibility, emulate the simple serializer using the contextual one
+        private byte[] EmulateSimpleSerializer(T msg)
         {
-            get
+            // TODO(jtattermusch): avoid the allocation by passing a thread-local instance
+            var context = new EmulatedSerializationContext();
+            this.contextualSerializer(msg, context);
+            return context.GetPayload();
+        }
+
+        // for backward compatibility, emulate the simple deserializer using the contextual one
+        private T EmulateSimpleDeserializer(byte[] payload)
+        {
+            // TODO(jtattermusch): avoid the allocation by passing a thread-local instance
+            var context = new EmulatedDeserializationContext(payload);
+            return this.contextualDeserializer(context);
+        }
+
+        // for backward compatibility, emulate the contextual serializer using the simple one
+        private void EmulateContextualSerializer(T message, SerializationContext context)
+        {
+            var payload = this.serializer(message);
+            context.Complete(payload);
+        }
+
+        // for backward compatibility, emulate the contextual deserializer using the simple one
+        private T EmulateContextualDeserializer(DeserializationContext context)
+        {
+            return this.deserializer(context.PayloadAsNewBuffer());
+        }
+
+        internal class EmulatedSerializationContext : SerializationContext
+        {
+            bool isComplete;
+            byte[] payload;
+
+            public override void Complete(byte[] payload)
+            {
+                GrpcPreconditions.CheckState(!isComplete);
+                this.isComplete = true;
+                this.payload = payload;
+            }
+
+            internal byte[] GetPayload()
+            {
+                return this.payload;
+            }
+        }
+
+        internal class EmulatedDeserializationContext : DeserializationContext
+        {
+            readonly byte[] payload;
+
+            public EmulatedDeserializationContext(byte[] payload)
+            {
+                this.payload = GrpcPreconditions.CheckNotNull(payload);
+            }
+
+            public override int PayloadLength => payload.Length;
+
+            public override byte[] PayloadAsNewBuffer()
```

Comment:
Payload is not newly allocated here. Could this break some imaginary `contextualDeserializer` implementations that rely on the the "newly allocated" part of the API

Reply:
Good point. I added a doc comment explaining that exotic deserializers that require to call PayloadAsNewBuffer() multiple times are not supported (there's no good reason why one should invoke that method twice). https://github.com/grpc/grpc/pull/16367/files#diff-22501e836d94b70164ce50d8b0c0c4e5R40

Reply:
see comment below, it might help to prevent the assumption with an assertion (would also help readability)